Add validation util class for validating webhook

doxiao opened the pull request:
This PR contains an utility class for validating webhook. The initial version only covers the checks for replicas.

russgold commented on operator/src/main/java/oracle/kubernetes/operator/utils/ValidationUtils.java:
```diff
@@ -0,0 +1,137 @@
+   * @param proposedDomain domain that needs to be validated.
+   * @return true if valid, otherwise false
+   */
+  public static boolean validateDomain(@NotNull Domain existingDomain, @NotNull Domain proposedDomain) {
```
The name is a bit confusing: it's not actually "validating" a domain, is it? I infer that it is actually returning true if the proposed change is valid. If so, it would probably be better to call it something like, `isPermittedChange` or `isDomainReplacementAllowed`. Methods that return booleans are conventionally given names that are predicates.

russgold commented on operator/src/main/java/oracle/kubernetes/operator/utils/ValidationUtils.java:
```diff
@@ -0,0 +1,137 @@
+        || shouldIntrospect(existingDomain, proposedDomain);
+  }
+
+  private static boolean unchanged(@NotNull Domain existingDomain, @NotNull Domain proposedDomain) {
```
`isUnchanged` would be a more conventional name for this.

russgold commented on operator/src/main/java/oracle/kubernetes/operator/utils/ValidationUtils.java:
```diff
@@ -0,0 +1,137 @@
+        if (noAuxiliaryImagesConfigured(existingDomain) && noAuxiliaryImagesConfigured(proposedDomain)) {
+          return !Objects.equals(getImage(existingDomain), getImage(proposedDomain));
+        } else {
+          return noAuxiliaryImagesConfigured(existingDomain)
```
This block could probably be extracted into its own method, which could be given a descriptive name to explain what is going on, such as `areAuxiliaryImagesChanged`. That would make the top-level method more clear.

russgold commented on operator/src/main/java/oracle/kubernetes/operator/utils/ValidationUtils.java:
```diff
@@ -0,0 +1,137 @@
+  private static boolean validReplicas(@NotNull Domain domain) {
+    return Optional.of(domain)
+        .map(Domain::getStatus)
+        .map(DomainStatus::getClusters)
```
I have found that it is easier to understand this sequence of steps when they are grouped on a single line, as they represent a single concept rather than multiple transformations:

```
.map(DomainStatus::getClusters).orElse(Collections.emptyList()).stream()
```

Alternately, you could extract the three steps into a single method named something like `getClustersAsStream`

russgold commented on operator/src/main/java/oracle/kubernetes/operator/utils/ValidationUtils.java:
```diff
@@ -0,0 +1,137 @@
+    return domain.getAuxiliaryImages() == null;
+  }
+
+  private static boolean validReplicas(@NotNull Domain domain) {
```
The name of this method and the next are unclear. It seems to me that it returns true if all clusters in the domain have valid replica counts. If so, a better name for this one might be `allReplicaCountsValid` and for the next `isReplicaCountValid`.

russgold reviewed (commented):
Overall looks very good. I have identified some places where the readability could be improved.

doxiao commented:
> Overall looks very good. I have identified some places where the readability could be improved.

I have addressed all of your comments. Please review again. thanks.
